src/BenchmarkDotNet.Core/Engines/GcStats.cs

```diff
@@ -0,0 +1,120 @@
+        {
+            return new GcStats(
+                left.Gen0Collections + right.Gen0Collections,
+                left.Gen1Collections + right.Gen1Collections,
+                left.Gen2Collections + right.Gen2Collections,
+                left.AllocatedBytes + right.AllocatedBytes,
+                left.TotalOperations + right.TotalOperations);
+        }
+
+        public static GcStats operator -(GcStats left, GcStats right)
+        {
+            return new GcStats(
+                Math.Max(0, left.Gen0Collections - right.Gen0Collections),
+                Math.Max(0, left.Gen1Collections - right.Gen1Collections),
+                Math.Max(0, left.Gen2Collections - right.Gen2Collections),
+                Math.Max(0, left.AllocatedBytes - right.AllocatedBytes),
+                Math.Max(0, left.TotalOperations - right.TotalOperations));
+        }
+
+        public GcStats WithTotalOperations(long totalOperationsCount)
+            => this + new GcStats(0, 0, 0, 0, totalOperationsCount);
+
+        internal static GcStats ReadInitial(bool isDiagnosticsEnabled)
+        {
+            // this might force GC.Collect, so we want to do this before collecting collections counts
+            long allocatedBytes = GetAllocatedBytes(isDiagnosticsEnabled); 
+
+            return new GcStats(
+                GC.CollectionCount(0),
+                GC.CollectionCount(1),
+                GC.CollectionCount(2),
+                allocatedBytes,
+                0);
+        }
+
+        internal static GcStats ReadFinal(bool isDiagnosticsEnabled)
+        {
+            return new GcStats(
+                GC.CollectionCount(0),
+                GC.CollectionCount(1),
+                GC.CollectionCount(2),
+
+                // this might force GC.Collect, so we want to do this after collecting collections counts 
+                // to exclude this single full forced collection from results
+                GetAllocatedBytes(isDiagnosticsEnabled), 
+                0);
+        }
+
+        public static GcStats FromForced(int forcedFullGarbageCollections)
+            => new GcStats(forcedFullGarbageCollections, forcedFullGarbageCollections, forcedFullGarbageCollections, 0, 0);
+
+        private static long GetAllocatedBytes(bool isDiagnosticsEnabled)
+        {
+#if NETCOREAPP11 // when MS releases new version of .NET Runtime to nuget.org
+            return GC.GetAllocatedBytesForCurrentThread(); // https://github.com/dotnet/corefx/pull/12489
+#elif CLASSIC
+            if (!isDiagnosticsEnabled)
+                return 0;
+
+            // "This instance Int64 property returns the number of bytes that have been allocated by a specific 
+            // AppDomain. The number is accurate as of the last garbage collection." - CLR via C#
+            // so we enforce GC.Collect here just to make sure we get accurate results
+            GC.Collect();
+            return AppDomain.CurrentDomain.MonitoringTotalAllocatedMemorySize;
```

Comment:
This will throw a [NIE](https://github.com/mono/mono/blob/7ca51073c638becc090e2c974ecaa739af62e4a3/mcs/class/corlib/System/AppDomain.cs#L1663-L1665) for Mono.